contrib/babelfishpg_tsql/src/linked_servers.c

```diff
@@ -1317,6 +1384,528 @@ openquery_imp(PG_FUNCTION_ARGS)
 	return (Datum) 0;
 }
 
+/*
+ * Map PostgreSQL/Babelfish type OID to TDS type for RPC parameter binding
+ */
+int
+get_tds_type_from_pg_oid(Oid pgtype)
+{
+	/* First check for Babelfish-specific types */
+	if (common_utility_plugin_ptr && common_utility_plugin_ptr->lookup_tsql_datatype_oid)
+	{
+		Oid varchar_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("varchar");
+		Oid nvarchar_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("nvarchar");
+		Oid char_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("char");
+		Oid nchar_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("nchar");
+		Oid int_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("int");
+		Oid bigint_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("bigint");
+		Oid tinyint_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("tinyint");
+		Oid datetime_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("datetime");
+		Oid smalldatetime_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("smalldatetime");
+		Oid datetime2_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("datetime2");
+		Oid float_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("float");
+		Oid real_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("real");
+		Oid varbinary_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("varbinary");
+		Oid binary_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("binary");
+		Oid uniqueidentifier_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("uniqueidentifier");
+		Oid bit_oid = (*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("bit");
+		
+		if (pgtype == varchar_oid)
+			return SYBVARCHAR;
```

Comment:
Please avoid referencing direct FreeTDS type macros. Instead use or define macros for these in linked_server.h . This is necessary to minimise code change if we ever decide to use another client library.